app/ui.js

```diff
@@ -939,54 +938,179 @@ const UI = {
     },
 
     readClipboard(callback) {
-      var readfuture = navigator.clipboard.readText();
-      readfuture
-        .then(text => callback(text))
-        .catch(() =>
-            console.error("Failed to read contents of clipboard (trying to copy to VNC clipboard)")
-        );
+	var readfuture = navigator.clipboard.readText();
+	readfuture
+            .then(text => callback(text))
+            .catch(() =>
+		   Log.Debug("Failed to read system clipboard")
+		  );
     },
 
+    // Copy text from NoVNC desktop to local computer
     clipboardReceive(e) {
-        Log.Debug(">> UI.clipboardReceive: " + e.detail.text.substr(0, 40) + "...");
-        document.getElementById('noVNC_clipboard_text').value = e.detail.text;
-        Log.Debug("<< UI.clipboardReceive");
+	var curvalue = document.getElementById('noVNC_clipboard_text').value;
+	if (curvalue != e.detail.text) {
+            Log.Debug(">> UI.clipboardReceive: " + e.detail.text.substr(0, 40) + "...");
+            document.getElementById('noVNC_clipboard_text').value = e.detail.text;
+            Log.Debug("<< UI.clipboardReceive");
+
+            var writefuture = navigator.clipboard.writeText(e.detail.text);
+            writefuture
+		.then(function() {
+                    /* clipboard successfully set */
+		    UI.popupMessage("Selection Copied");
+		}, function() {
+                    console.error("Failed to write system clipboard (trying to copy from NoVNC clipboard)")
+		});
+	}
+    },
+
+    popupMessage(msg) {
+	// Quick popup to give feedback that selection was copied
+	setTimeout(UI.showOverlay.bind(this, msg, 500), 200);
+    },
+
+    // Enter and focus events come when we return to NoVNC.
+    // In both cases, check the local clipboard to see if it changed.
+    focusVNC() {
+	UI.copyFromLocalClipboard();
+    },
+    enterVNC() {
+	UI.copyFromLocalClipboard();
+    },
+    copyFromLocalClipboard() {
+	UI.readClipboard(text => {
+	    var clipVal = document.getElementById('noVNC_clipboard_text').value;
+	    if ( clipVal != text ) {
```

Comment:
I think this is essential, otherwise we have no mechanism that avoids just blindly overwriting the server clipboard whenever focus changes.

However this is still a hack, so I think we should reference the `clipboardchange` event and that we're waiting for browsers to implement it.